### DISCUSSION TITLE
In case of fallback: true, does the getStaticPaths() generate additional paths before data is fetched in getStaticProps() and returned as props?

### DISCUSSION THREAD

**Discussion by edgarasben** [vercel/next.js#40969] [Help] [answered]
In case of fallback: true, does the `getStaticPaths()` generate additional paths before data is fetched in `getStaticProps()` and returned as `props`?

I want to make sure that additional static pages are not generated on a request if certain slugs are not found in a database.

**Comment by icyJoseph** [comment, marked as answer]
`getStaticPaths` is only called at build time, as per [docs](https://nextjs.org/docs/basic-features/data-fetching/get-static-paths#when-does-getstaticpaths-run):

> getStaticPaths will only run during build in production, it will not be called during runtime. You can validate code written inside getStaticPaths is removed from the client-side bundle [with this tool](https://next-code-elimination.vercel.app/).

Of course in `dev` mode the above changes:

> In development (next dev), getStaticPaths will be called on every request.

**Reply by edgarasben** [reply]
I am referring to `fallback: true` case, when static pages are generated on demand when users visit a page.

For example, I generate `/blog/1` at built time, but then `/blog/2` can be generated on request. However, I want to prevent generating `/blog/3` is it doesn't exist. Is it enough to return `notFound` in getStaticProps to prevent this?

**Reply by icyJoseph** [reply]
Not in `getStaticPaths`, but you can check that directly on `getStaticProps` and return `{ notFound: true}` there.

I can post a snippet later.

**Reply by edgarasben** [reply]
Ok, so before `{ notFound: true}` is returned in `getStaticProps` the static page is not generated in the background for sure?

**Reply by icyJoseph** [reply]
Mmm I am worried we might be talking about various things. 

Let me expand a bit. First let's look at this code:

```tsx
import type { GetStaticPaths, GetStaticProps, NextPage } from "next";

import Link from "next/link";
import Router, { useRouter } from "next/router";
import { useEffect } from "react";

export const PostsPage: NextPage = () => {
  const { query, isFallback } = useRouter();
  const { page } = query;
  const numPage = Number(page);

  useEffect(() => {
    if (!isFallback && isNaN(numPage)) {
      Router.reload();
    }
  }, [numPage, isFallback]);

  if (isFallback) {
    return <div>Loading...</div>;
  }

  return (
    <div>
      Post page: {page} from a total
      <nav>
        <ul>
          {[-1, 1].map((step) => {
            const index = numPage + step;

            return (
              <li key={index}>
                <Link href={`/posts/${index}`}>
                  <a>Nav to: {index}</a>
                </Link>
              </li>
            );
          })}
        </ul>
      </nav>
    </div>
  );
};

/* get this from your CMS */
const paths = [
  { params: { page: "0" } },
  { params: { page: "1" } },
  { params: { page: "2" } },
  { params: { page: "3" } },
];

export const getStaticPaths: GetStaticPaths<{ page: string }> = () => {
  return {
    paths,
    fallback: true,
  };
};

export const getStaticProps: GetStaticProps = async (ctx) => {
  const page = ctx.params?.page;
  // returning revalidate in both cases here, purely for example sake!
  if (typeof page === "undefined") return { notFound: true, revalidate: 1_000 };
  if (Array.isArray(page)) return { notFound: true, revalidate: 1_000 };

  const asNumber = Number(page);
  // same here, including the revalidate for demonstration only
  if (isNaN(asNumber)) return { notFound: true, revalidate: 1_000 };

  return {
    props: {},
    revalidate: 1_000,
  };
};

export default PostsPage;
```

There's quite a bit to unpack there.

The one thing to notice is that, we do not want to render `/posts/foo`, or anything that can't be parsed as a `Number`. We put in place a few `isNaN` checks for `Number`.

Also notice that I am returning `revalidate` even from the `notFound` cases. In this case, this is not needed, things that can't be parsed as number will always fail, but in a real application, perhaps a post is not available yet, but it will be in the future. If we don't return revalidate the `notFound: true` result includes a header `Cache-Control: s-maxage=31536000, stale-while-revalidate`, which means we are in trouble for one year. 

The above is also not every straightforward. Because the `props` result also has `revalidate`, when we arrive to a page that doesn't exist `in-memory`, the cache header would be `Cache-Control: s-maxage=1000, stale-while-revalidate`. Even if the `isNaN` not found branch returns no revalidate, but it'd be so, only for that one browser that requested it. If I switch to Firefox and request the faulty page, `/posts/foo`, I'd get the 1 year cache. 

Alright, almost none of the above applies to your question though, but it is important to be aware of these things.

Let's now look at the `.next/server/pages/posts/` folder where my application is running from, after it's build. 

```shell
0.html
0.json
1.html
1.json
2.html
2.json
3.html
3.json
[page].html
[page].js
[page].js.nft.json
```

It contains the four pages I generate in `getStaticPaths`, now let me start the server and navigate to `/posts/4`:

```shell
0.html
0.json
1.html
1.json
2.html
2.json
3.html
3.json
4.html
4.json
5.html
5.json
[page].html
[page].js
[page].js.nft.json
```
 
We see that data for page 4 has been created, as well as data for page 5, which `next/link` automatically prefetches for us, it did the same for page 3, but it already existed.

Now let's go to `/posts/foo`, again server just started and we navigated to `/posts/4`.

First of all, since the page did not exist, the `Page Fallback` UI kicks in, we can see this through the `isFallback` property in the router. 

The browser renders `Loading...`, when `isFallback` switches back to `false`, the `useEffect` I put in the client forced a page reload, and we show a **_404_**. Blink and you'll miss it!

Now from another tab, or through refresh reload,  another browser, etc, let's go back to `/posts/foo`. We get a 404 directly, no fallback UI what so ever. Because `in-memory` it is know that this should just be a `404`. 

And if we check `.next/server/pages/posts/`, you'd see that nothing has changed.

Restarting the server keeps the generated pages 4 and 5, of course 0 to 3 are also preserver in `.next/server/pages/posts/`, but going to `/posts/foo` triggers the UI (`useEffect` + `reload`) logic again, which I why I reckon it is saved on the server memory.

Hope I didn't confuse you even more, and that at least I covered what you wanted.

**Reply by edgarasben** [reply]
This was really helpful guide Joseph! I think you could turn this to a blog post, as I really never hear about how this lifecycle works in such detail. I am not even sure NextJS docs has these parts.

And, thank you so much for making me aware of the revalidation needed ALSO for notFound cases! 

Now I understand better how 404 pages are first created and cached. 

I am wondering though, let's say by accident a 404 as cached to CDN to 1 year as you say. Do you know when/how the CDN cache be purged? Sounds a bit crazy that every 404 page is cached to a CDN for so long. I haven't encountered this issue yet to be honest.

**Reply by icyJoseph** [reply]
It wouldn't be cached to a CDN, necessarily, but rather to the browser. If there's a CDN layer, then yeah, maybe that layer will try to honour the cache-header. 

It is not a very common issue I guess, but it is good to be aware of it, perhaps you have posts that do not yet exist, and returning a one year 404 on them is not very nice.

Don't forget to mark as answered!

**Reply by edgarasben** [reply]
@icyJoseph yup, now I am more aware of 404 caching. Vercel says they wipe cache every 31 days 

> Static files are cached for up to 31 days.

Thanks for your time again!
